### PR TITLE
Fixes #32488 - use rpm to remove pulp2 packages

### PR DESCRIPTION
**Pull request by jlsherrill**
as yum can remove other packages unexpectedly
(such as mod_ssl and more)

### CODE REVIEW

File: definitions/procedures/pulp/remove.rb [hunk 1]
```diff
@@ -23,16 +23,22 @@ def pulp_data_dirs
...
+                        pulp-rpm-handlers pulp-rpm-plugins pulp-rpm-yumplugins pulp-selinux pulp-server python-bson python-crane
+                        python-isodate python-mongoengine python-nectar python-pulp-agent-lib python-pulp-bindings python-pulp-client-lib
+                        python-pulp-common python-pulp-deb-common python-pulp-devel python-pulp-docker-common python-pulp-integrity
+                        python-pulp-manifest python-pulp-oid_validation python-pulp-ostree-common python-pulp-puppet-common
+                        python-pulp-python-common python-pulp-repoauth python-pulp-rpm-common python-pulp-streamer python-pymongo
+                        python-pymongo-gridfs python2-amqp python2-billiard python2-celery python2-debpkgr python2-django python2-kombu
+                        python2-solv python2-vine)
+
+      @installed_pulp_packages ||= all_possible.select { |pkg| find_package(pkg) }
+      @installed_pulp_packages
```
Comment (jlsherrill) [inline definitions/procedures/pulp/remove.rb:41]: here we only consider packages actually installed, and cache to not have to recalculate again when this method is called a 2nd time.

File: definitions/procedures/pulp/remove.rb [hunk 2]
```diff
@@ -22,17 +22,29 @@ def pulp_data_dirs
...
-      ]
+      possible = %w[pulp-admin-client pulp-agent pulp-consumer-client pulp-deb-admin-extensions
+                    pulp-deb-plugins pulp-docker-admin-extensions pulp-docker-plugins
+                    pulp-nodes-admin-extensions pulp-nodes-child pulp-nodes-common
+                    pulp-nodes-consumer-extensions pulp-nodes-parent pulp-ostree-admin-extensions
+                    pulp-ostree-plugins pulp-puppet-admin-extensions
+                    pulp-puppet-consumer-extensions pulp-puppet-handlers pulp-puppet-plugins
+                    pulp-puppet-tools pulp-python-admin-extensions pulp-python-plugins
+                    pulp-rpm-admin-extensions pulp-rpm-consumer-extensions pulp-rpm-handlers
+                    pulp-rpm-plugins pulp-rpm-yumplugins pulp-selinux pulp-server python-bson
```
Comment (ekohl) [inline definitions/procedures/pulp/remove.rb:35]: Can't you use `rpm -qa | grep ^pulp-` or will that also match pulp 3 packages?

File: definitions/procedures/pulp/remove.rb [hunk 3]
```diff
@@ -22,17 +22,29 @@ def pulp_data_dirs
...
+                    python-crane python-isodate python-mongoengine python-nectar
+                    python-pulp-agent-lib python-pulp-bindings python-pulp-client-lib
+                    python-pulp-common python-pulp-deb-common python-pulp-devel
+                    python-pulp-docker-common python-pulp-integrity python-pulp-manifest
+                    python-pulp-oid_validation python-pulp-ostree-common python-pulp-puppet-common
+                    python-pulp-python-common python-pulp-repoauth python-pulp-rpm-common
+                    python-pulp-streamer python-pymongo python-pymongo-gridfs python2-amqp
```
Comment (ekohl) [inline definitions/procedures/pulp/remove.rb:42]: Same here with `^python-pulp-`

File: definitions/procedures/pulp/remove.rb (hunk 2 above)
Comment (jlsherrill) [reply definitions/procedures/pulp/remove.rb:35]: I think both these suggestions would work, but to be honest i prefer this explicit list especially given the side effects we saw from users.

**Comment by upadhyeammit** [comment]
Hello @jlsherrill 

Changes look good to me! Apart from these changes can you also modify `data_dir_removal_cmds` as,
~~~
    def data_dir_removal_cmds
      pulp_data_dirs.collect { |dir| "rm -rf #{dir}" if File.directory?(dir)}.compact
    end
~~~

Accordingly remove directory check from `delete_pulp_data`. This way we will only show the directories we will actually remove and avoid the split in `delete_pulp_data`.

Thank you

**Comment by jlsherrill** [comment]
updated!

**Comment by upadhyeammit** [comment]
> updated!

Hello @jlsherrill can you remove `if File.directory?(cmd.split[2])` from `delete_pulp_data` as we already have checked this in `data_dir_removal_cmds`. Thank you!

**Comment by jlsherrill** [comment]
updated!

**Review by upadhyeammit — APPROVED** [review]

**Comment by jturel** [comment]
Hit this while testing on a 4.0 box I upgrade from 3.18:

```
foreman-maintain content remove-pulp2
Running Remove Pulp2 and mongodb packages and data
================================================================================
Remove pulp2: 

WARNING: All pulp2 packages will be removed with the following commands:

# rpm -e pulp-deb-plugins  pulp-docker-plugins  pulp-puppet-plugins  pulp-puppet-tools  pulp-rpm-plugins  pulp-selinux  pulp-server  python-bson  python-crane  python-isodate  python-mongoengine  python-nectar  python-pulp-common  python-pulp-deb-common  python-pulp-docker-common  python-pulp-oid_validation  python-pulp-puppet-common  python-pulp-repoauth  python-pulp-rpm-common  python-pulp-streamer  python-pymongo  python-pymongo-gridfs  python2-amqp  python2-billiard  python2-celery  python2-debpkgr  python2-django  python2-kombu  python2-solv  python2-vine
# yum remove rh-mongodb34-*

All pulp2 data will be removed.

# rm -rf /var/lib/pulp/published
# rm -rf /var/lib/pulp/uploads
# rm -rf /var/lib/mongodb/

Do you want to proceed?, [y(yes), q(quit)] y
/ Removing pulp2 packages                                             [FAIL]    
Failed executing rpm -e pulp-deb-plugins  pulp-docker-plugins  pulp-puppet-plugins  pulp-puppet-tools  pulp-rpm-plugins  pulp-selinux  pulp-server  python-bson  python-crane  python-isodate  python-mongoengine  python-nectar  python-pulp-common  python-pulp-deb-common  python-pulp-docker-common  python-pulp-oid_validation  python-pulp-puppet-common  python-pulp-repoauth  python-pulp-rpm-common  python-pulp-streamer  python-pymongo  python-pymongo-gridfs  python2-amqp  python2-billiard  python2-celery  python2-debpkgr  python2-django  python2-kombu  python2-solv  python2-vine, exit status 30:
 error: Failed dependencies:
        pulp-server >= 2.8 is needed by (installed) pulp-katello-1.0.3-1.el7.noarch
        python-pulp-rpm-common is needed by (installed) pulp-katello-1.0.3-1.el7.noarch
--------------------------------------------------------------------------------
Scenario [Remove Pulp2 and mongodb packages and data] failed.

The following steps ended up in failing state:

  [pulp-remove]

Resolve the failed steps and rerun
the command. In case the failures are false positives,
use --whitelist="pulp-remove"
```

**Comment by jturel** [comment]
I added pulp-katello to the rpms to be removed and it succeeded.

Here's what's left over containing "pulp" or "mongo":

```
pulp-client-1.0-1.noarch
pulpcore-selinux-1.2.4-1.el7.x86_64
python3-mongoengine-0.20.0-1.el7.noarch
python3-pulp-2to3-migration-0.11.1-1.el7.noarch
python3-pulp-certguard-1.1.0-1.el7.noarch
python3-pulp-container-2.2.2-1.el7.noarch
python3-pulpcore-3.9.1-2.el7.noarch
python3-pulp-deb-2.9.2-1.el7.noarch
python3-pulp-file-1.5.0-1.el7.noarch
python3-pulp-rpm-3.10.0-1.el7.noarch
python3-pymongo-3.11.0-1.el7.x86_64
tfm-rubygem-pulp_2to3_migration_client-0.7.0-1.el7.noarch
tfm-rubygem-pulp_ansible_client-0.6.0-1.el7.noarch
tfm-rubygem-pulp_certguard_client-1.0.3-1.el7.noarch
tfm-rubygem-pulp_container_client-2.2.0-1.el7.noarch
tfm-rubygem-pulpcore_client-3.9.0-1.el7.noarch
tfm-rubygem-pulp_deb_client-2.9.1-1.el7.noarch
tfm-rubygem-pulp_file_client-1.5.0-1.el7.noarch
tfm-rubygem-pulp_rpm_client-3.10.0-1.el7.noarch
tfm-rubygem-smart_proxy_pulp-2.1.0-3.fm2_2.el7.noarch
```

The leftovers seem valid!

**Comment by jlsherrill** [comment]
not sure why i didn't hit that, but easy enough to add!  Thanks @jturel   Added!

**Review by jturel — APPROVED** [review]
ACK - tests well!

**Comment by jlsherrill** [comment]
@upadhyeammit can you merge please :)

**Pull request merged**